[CORE-6875] `cloud_storage`: fall back to main manifest search in `async_manifest_view::get_term_last_offset()`

WillemKauf opened the pull request:
When we fail to find an entry in the spillover map via `get_spillover_upper_bound_by_term()`, we need to fall back to searching the main manifest for the next `segment` term to properly respond to a `OffsetForLeaderEpoch` query.

Consider the following edge case:

Populate 2 spillover manifests in the `partition_manifest`'s `spillover_map`. The desired term 1, is last entry in the spillover manifests, i.e there are no higher term entries in `get_segment_term_column`, and the main manifest contains no segments with a lower term than the desired term. This can be illustrated as:

```
Main manifest: [2, 3]
Spillover map: [[0], [1]]
```

`get_spillover_upper_bound_by_term()` will return `std::nullopt`, since no higher bound for the desired term 1 was found in the 2 entries of the spillover map. Previously, this case would erroneously be treated as an early return of `std::nullopt` from `get_term_last_offset()`. However, we still need to consider the main manifest in this case, as the base offset of the `segment` with term 2 present in the main manifest will give us our proper response to the query.

This bug would manifest as an incorrect response to `OffsetForLeaderEpoch`, which could lead to reset consumers.

https://redpandadata.atlassian.net/browse/CORE-6875 is the JIRA ticket for this failure which also has a comment which shows the consumer side logs.

## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v25.2.x
- [X] v25.1.x
- [X] v24.3.x
- [ ] v24.2.x

## Release Notes


### Bug Fixes

* Fixes a bug in which a `partition` for a tiered storage enabled `topic` could provide an incorrect response to an `OffsetForLeaderEpoch` request

oleiman commented on src/v/cloud_storage/async_manifest_view.cc:
```diff
@@ -1700,23 +1700,32 @@ async_manifest_view::materialize_manifest(
     const auto& last_term_col = spillover_map.get_segment_term_column();
     size_t sp_index = 0;
+    bool found_term = false;
     for (auto last_term : last_term_col) {
         if (last_term > term()) {
             break;
         }
+        if (last_term == term()) {
+            found_term = true;
+        }
         sp_index++;
     }
 
     if (sp_index == last_term_col.size()) {
-        return std::nullopt;
+        if (found_term) {
+            // The upper bound was the last entry in the column.
+            return sp_index - 1;
+        } else {
+            return std::nullopt;
+        }
     }
```
nitpick: sort of hard to follow. what about

```suggestion
    size_t idx = 0;
    std::optional<size_t> result;
    for (auto last_term : last_term_col) {
        if (last_term > term()) {
            break;
        } 
        if (last_term == term()) {
            result = idx;
        }
        ++idx;
    }
    return result;
```

WillemKauf replied:
Thanks, took this snippet, I agree it's better, though the code needs to be:

```cpp
    size_t idx = 0;
    std::optional<size_t> result = std::nullopt;
    // We are searching for the spillover manifest that has a
    // `segment_term_column` entry equal to or higher than the desired term.
    for (auto last_term : last_term_col) {
        if (last_term >= term()) {
            result = idx;
        }
        if (last_term > term()) {
            break;
        }
        ++idx;
    }
    return result;
```

oleiman replied:
too right 👍

oleiman replied:
actually, if that's what the function is supposed to do, isn't it, in fact, an upper_bound? and does that jibe w/ the PR comment?

Lazin commented:
I'm not sure that this is the bug. If all spillover manifests have segment term (which is a largest term in the spillover manifest) smaller than the query then the search should return nullopt. After that we should continue searching the main manifest.

WillemKauf commented:
> After that we should continue searching the main manifest.

Huh. That wasn't my mental model but I suppose it makes sense. 

I suppose the early return in `get_term_last_offset()` should instead be a query on the main manifest instead then.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--9b518881-6b24-47bb-a324-0c6502bd2da6-->
<!--TEST RESULTS-->
<!--f030ef2c-f32f-4aa0-9546-cc341472b403-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69132">build#69132</a></summary>


| test_class               | test_method                           | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:--------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MasterTestSuite          | test_chunk_dl_with_random_http_errors |                                                                                                                                                 | unit        | https://buildkite.com/redpanda/redpanda/builds/69132#0198151f-e227-48ea-928c-6acddf16a99b | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest | test_node_operations                  | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69132#01981537-d68e-46ba-8d1a-6fb828c90ae7 | FLAKY         | 20/21    | upstream reliability is '99.62825278810409'. current run reliability is '95.23809523809523'. drift is 4.39016 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69188">build#69188</a></summary>


| test_class      | test_method             | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------|:------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ArchivalTest    | test_connection_flicker | {"cloud_storage_type": 2} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69188#01981923-d3bb-40df-b11f-3c921a5a887e | FLAKY         | 20/21    | upstream reliability is '99.21465968586386'. current run reliability is '95.23809523809523'. drift is 3.97656 and the allowed drift is set to 50. The test should PASS |
| MaintenanceTest | test_maintenance_sticky | {"use_rpk": true}         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69188#0198191f-8a4d-4a7b-aec1-2307b698ef65 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<!--/body-->

WillemKauf committed:
`cloud_storage`: fall back to main manifest search in `get_term_last_offset()`

When we fail to find an entry in the spillover map via
`get_spillover_upper_bound_by_term()`, we need to fall back to searching
the main manifest for the next `segment` term to properly respond to
a `OffsetForLeaderEpoch` query.

Consider the following edge case:

Populate 2 spillover manifests in the `partition_manifest`'s `spillover_map`.
The desired term 1, is last entry in the spillover manifests, i.e there are
no higher term entries in `get_segment_term_column`, and the main manifest
contains no segments with a lower term than the desired term. This can be
illustrated as:

```
Main manifest: [2, 3]
Spillover map: [[0], [1]]
```

`get_spillover_upper_bound_by_term()` will return `std::nullopt`, since no
higher bound for the desired term 1 was found in the 2 entries of the
spillover map. Previously, this case would erroneously be treated as an
early return of `std::nullopt`. However, we still need to consider the main
manifest in this case, as the base offset of the `segment` with term 2
present in the main manifest will give us our proper response to the query.

WillemKauf commented:
> I'm not sure that this is the bug. If all spillover manifests have segment term (which is a largest term in the spillover manifest) smaller than the query then the search should return nullopt. After that we should continue searching the main manifest.

After grokking this whole code path better I agree. I have repushed to this PR with a completely different (and simpler) fix. Please LMKWYT

Lazin commented on src/v/cloud_storage/async_manifest_view.cc:
```diff
@@ -743,6 +743,12 @@ async_manifest_view::get_term_last_offset(model::term_id term) noexcept {
         const auto spill_index = get_spillover_upper_bound_by_term(term);
         if (!spill_index.has_value()) {
+            // Fall back to searching the main manifest for term.
+            for (const auto& p : stmm) {
```
I don't think we need to iterate here. On line 723 we're checking `stmm.begin()->segment_term <= term`. If this is not the case we're going into this code path. So the invariant here is that `stmm.begin()->segment_term > term`.

Also, the cloud storage log may have gaps so we shouldn't expect to see `term() + 1` here. We may see `term() + 2` or `+10` or whatever. This breaks some api guarantees but we can't really do anything here if TS was disable for long enough and then re-enabled.

So basically, I think it's enough to just return a base offset of the last segment - 1 here without going through the manifest. It's also guaranteed that the main manifest (or STM manifest) is not empty because you have spillover manifests and this code is not async.

WillemKauf replied:
> Also, the cloud storage log may have gaps so we shouldn't expect to see term() + 1 here. We may see term() + 2 or +10 or whatever. This breaks some api guarantees but we can't really do anything here if TS was disable for long enough and then re-enabled.

Right yeah, that's what I was worried about here, I wasn't sure whether an incorrect return value or an empty return value would be better in this case, which is why I put the check for specifically `term() + 1`.

> So basically, I think it's enough to just return a base offset of the last segment - 1 here without going through the manifest.

Why would the `base_offset - 1` of the last segment be the right response? What about the case

```
Main manifest: [4, 5]
Spillover map: [[0], [1], [2], [3]]
```

with desired term = 3? That would lead to an incorrect result

Lazin replied:
shouldn't we find the spillover manifest [3] in this case?

WillemKauf replied:
> shouldn't we find the spillover manifest [3] in this case?

No, because `get_spillover_upper_bound_by_term()` will return `std::nullopt`, because the logic in that function is looking for a higher term in the spillover manifests as its return condition in the happy path.

When the term boundaries are as such (i.e desired term is last one in the spillover map, and the higher term is only in the main manifest), that is the buggy edge case in the current code.

Lazin replied:
* we first check the condition on line 723, 4 <= 3 is false, so we're going for the spillover part
* spillover search in `get_spillover_upper_bound_by_term()` returns `nullopt` because everything there is smaller or equal to 3
* now we need to take the first segment in the manifest, which has term = 4 get its base_offset and return base-1, it's guaranteed that the first segment in the STM has term which is greater than 3 because of the first check and the fact that `get_spillover_upper_bound_by_term()` returns nullopt guarantees that the spillover part has segments with terms less or equal to 3

does it make sense?

WillemKauf replied:
Yep, it does. Pushed

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x
